tracker/server.go

```diff
@@ -0,0 +1,237 @@
+		hlog.URLHandler("url"),
+		hlog.MethodHandler("method"),
+		hlog.ProtoHandler("protocol"),
+		hlog.AccessHandler(zerologLoggerFunc),
+	)
+	r.Post("/listener_joined", ListenerAdd(ctx, recorder))
+	r.Post("/listener_left", ListenerRemove(ctx, recorder))
+
+	var err error
+
+	// setup the GRPC server that the rest will be poking
+	srv.rpcSrv = NewGRPCServer(ctx, recorder)
+	// and a listener for the GRPC server
+	srv.rpcLn, err = net.Listen("tcp", cfg.Conf().Tracker.RPCAddr)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	srv.httpSrv = &http.Server{
+		Addr:         cfg.Conf().Tracker.ListenAddr,
+		Handler:      r,
+		ReadTimeout:  time.Second * 10,
+		WriteTimeout: time.Second * 5,
+		BaseContext:  func(l net.Listener) context.Context { return ctx },
+	}
+	srv.httpLn, err = net.Listen("tcp", cfg.Conf().Tracker.ListenAddr)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	return srv, nil
+}
+
+func (s *Server) Close() error {
+	s.rpcSrv.Stop()
+	return s.httpSrv.Close()
+}
+
+func (s *Server) Start(ctx context.Context) error {
+	logger := zerolog.Ctx(ctx)
+	errCh := make(chan error, 2)
+	go func() {
+		s.mu.Lock()
+		ln := s.rpcLn
+		s.mu.Unlock()
+		logger.Info().Str("address", ln.Addr().String()).Msg("tracker started listening [grpc]")
+		errCh <- s.rpcSrv.Serve(ln)
+	}()
+	go func() {
+		s.mu.Lock()
+		ln := s.httpLn
+		s.mu.Unlock()
+		logger.Info().Str("address", ln.Addr().String()).Msg("tracker started listening [http]")
+		errCh <- s.httpSrv.Serve(ln)
+	}()
+
+	select {
+	case <-ctx.Done():
+		return s.Close()
+	case err := <-errCh:
+		return err
+	}
+
+}
+
+func (srv *Server) handleResume(ctx context.Context, cfg config.Config) error {
+	parent, err := graceful.FD2Unix(3)
+	if err != nil {
+		return err
+	}
+	defer parent.Close()
+
+	return srv.readSelf(ctx, cfg, parent)
+}
+
+func (srv *Server) handleRestart(ctx context.Context, cfg config.Config) error {
+	dst, err := graceful.StartChild()
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+
+	return srv.writeSelf(dst)
+}
+
+type httpWire string
+type rpcWire string
+
+// type wireServer struct {
+// 	HTTPServer, RPCServer string
+// }
+
+func (srv *Server) writeSelf(dst *net.UnixConn) error {
+	var hw httpWire
+	var rw rpcWire
+	rw = rpcWire(srv.cfg.Conf().Tracker.RPCAddr)
+	hw = httpWire(srv.cfg.Conf().Tracker.ListenAddr)
+
+	rpcFd, err := getFile(srv.rpcLn)
+	if err != nil {
+		return fmt.Errorf("fd failure in server: %w", err)
+	}
+	defer rpcFd.Close()
+
+	httpFd, err := getFile(srv.httpLn)
+	if err != nil {
+		return fmt.Errorf("fd failure in server: %w", err)
+	}
+	defer httpFd.Close()
+
+	err = graceful.WriteJSONFile(dst, rw, rpcFd)
+	if err != nil {
+		return err
+	}
+	err = graceful.WriteJSONFile(dst, hw, httpFd)
+	if err != nil {
+		return err
+	}
+
+	// TODO(kipu): serialize recorder
+	// return srv.recorder.writeSelf(dst)
+	return nil
+}
+
+func (srv *Server) readSelf(ctx context.Context, cfg config.Config, src *net.UnixConn) error {
```

Comment:
same goes for this function as `writeSelf`, make them two separate steps rather than interleaved